Add os grains test cases for Debian/Ubuntu and fix oscodename on Ubuntu

bdrung opened the pull request:
This pull requests adds os grains test cases for Debian and Ubuntu (to address comments in #34423) and fixes an incorrect oscodename grain on Ubuntu (found the test case).

bdrung committed:
Fix incorrect oscodename grain on Ubuntu

Use the full content of /etc/os-release for the Ubuntu 16.04 (artful)
os_grains test case. This reveals a bug that 'oscodename' is not
correctly calculated if /etc/os-release provides PRETTY_NAME:

FAIL: test_ubuntu_os_grains (tests.unit.grains.test_core.CoreGrainsTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/grains/test_core.py", line 439, in test_ubuntu_os_grains
    self._run_ubuntu_os_grains_tests(_os_release_map)
  File "tests/unit/grains/test_core.py", line 446, in _run_ubuntu_os_grains_tests
    self._run_os_grains_tests(os_release_map)
  File "tests/unit/grains/test_core.py", line 262, in _run_os_grains_tests
    self.assertEqual(os_grains.get('oscodename'), os_release_map['oscodename'])
AssertionError: 'Ubuntu 16.04.3 LTS' != 'xenial'
- Ubuntu 16.04.3 LTS
+ xenial

fixes #34423 for Ubuntu
Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
tests: Add Ubuntu 17.10 (artful) os_grains test case

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
tests: Deduplicate _run_os_grains_tests()

_run_suse_os_grains_tests() and _run_ubuntu_os_grains_tests() share most
of their logic. Combine the common part in _run_os_grains_tests() and
let the remaining small parts live in their origin functions.

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
tests: Add os_grains test cases for Debian

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
Fix osfinger grain on Debian

Rework _run_os_grains_tests() to check all os grains including osfinger.
Fix osfinger on Debian (e.g. use "Debian-9" instead of
"Debian GNU/Linux-9").

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
Make _parse_os_release() always callable

To simplify running unit tests against os_data(), make the
_parse_os_release() function always callable to avoid needing to mock
os.path.isfile().

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

isbm commented on tests/unit/grains/test_core.py:
```diff
@@ -243,10 +243,11 @@ def _import_mock(name, *args):
                                 with patch("salt.utils.fopen", mock_open()) as suse_release_file:
-                                    suse_release_file.return_value.__iter__.return_value = os_release_map.get('suse_release_file', '').splitlines()
+                                    suse_release_file.return_value.__iter__.return_value = \
+                                        os_release_map.get('suse_release_file', '').splitlines()
```
Doesn't looks like you've removed whole SUSE-related things from this function

bdrung replied:
`_run_suse_os_grains_tests` also uses this `_run_os_grains_tests` function. Besides that this code part was also part of `_run_ubuntu_os_grains_tests`.

isbm reviewed (changes requested):
I am not sure if this is a good approach to the unit tests, when you put pre-parsed data and simply comparing two dictionaries. It would make more sense to get a real data from Debian/Ubuntu machines, mock-up system reads and then pipe it through instead.

bdrung commented:
These are unit tests. Instead of using raw file content, we should add a unit tests for the `_parse_os_release` function.

bdrung commented:
Added unit tests for `_parse_os_release`.

isbm commented:
I mean, it would be nicer to have just a data samples somewhere (subdir?). I am sure this is different from distro to distro... Unfortunately.

bdrung commented:
Moving all the sample data into a subdirectory is an option (this could be done as additional commit / pull request)

isbm commented:
@bdrung since you're taking care of these unit tests, you could actually move that data right in this PR with no problems. It as a nice and complete unit test, so why not.

bdrung committed:
tests: Dissolve _run_ubuntu_os_grains_tests

_run_ubuntu_os_grains_tests is only used once and does not provide any
useful abstraction since the introduction of _run_os_grains_tests.

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung committed:
tests: Support reading os-release files from disk

Instead of storing pre-parsed os-release files, add support for reading
os-release files from disk.

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung commented:
Implemented as requested. The os-release files are placed in a os-releases directory.

isbm reviewed (approved):
Great!

rallytime commented:
@bdrung A small lint error here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/19475/violations/file/tests/unit/grains/test_core.py/

bdrung committed:
tests: Add unit tests for _parse_os_release()

Add tests cases to fully cover _parse_os_release().

Signed-off-by: Benjamin Drung <benjamin.drung@profitbricks.com>

bdrung commented:
Rebased and fixed pylint error.
